### PR TITLE
Amazon s3

### PR DESCRIPTION
**Pull request by laura-wenning**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,9 @@
     "test-coverage": "jest --coverage"
   },
   "dependencies": {
+    "@aws-sdk/client-s3": "^3.50.0",
+    "@aws-sdk/s3-request-presigner": "^3.50.0",
+    "@types/uuid": "^8.3.4",
```
Comment (laura-wenning) [inline package.json:16]: Move to dev dependencies

File: src/@owl-factory/aws/s3.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    Key: path,
+    ContentType: mimetype,
+  };
+
+  const command = new PutObjectCommand(params);
+  const uploadURL = await getSignedUrl(client, command, { expiresIn: URL_EXPIRATION_SECONDS });
+  return uploadURL;
+}
+
+async function getObjectMetadata(path: string) {
```
Comment (laura-wenning) [inline src/@owl-factory/aws/s3.ts:48]: Comment

File: src/@owl-factory/components/form/Input.tsx [hunk 3]
```diff
@@ -9,7 +9,7 @@ interface InputProps {
   onChange?: (e: React.ChangeEvent<any>) => void;
   placeholder?: string;
   style?: CSSProperties;
-  type: "text" | "password" | "email" | "number" | "textarea"
+  type: "text" | "password" | "email" | "number" | "textarea" | "file"
```
Comment (laura-wenning) [inline src/@owl-factory/components/form/Input.tsx:12]: Remove this. We need a future refactor of how we handle elements

File: src/controllers/data/FileData.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    // TODO - move into it's own function?
+    let res;
+    try {
+      res = await rest.put<{ file: FileDocument, uploadURL: string }>(`/api/files/begin-upload`, { doc });
+    } catch (e: any) {
+      // TODO - make a little more descriptive
+      throw "An error occured while attempting to reserve space for the file";
+    }
+
+    console.log(res.data)
```
Comment (laura-wenning) [inline src/controllers/data/FileData.ts:53]: Remove console.log

File: src/controllers/data/FileData.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    try {
+      awsRes = await s3.upload(res.data.uploadURL, values.file);
+    } catch (e: any) {
+      // We don't need to wait for this. If it fails, a cleanup action can take care of this
+      rest.delete(`/api/files/${res.data.file.ref}`, {});
+      throw "The file failed to upload successfully";
+    }
+
+    // Update File document with data
+    rest.post<{ file: FileDocument }>(`/api/files/validate-upload`, { file: res.data.file });
```
Comment (laura-wenning) [inline src/controllers/data/FileData.ts:66]: Add .then() for setting the new user and file documents

File: src/pages/api/files/validate-upload.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  // If we're not pending, then there is nothing to be done. Exit.
+  if (!fileDoc.isPending) {
+    this.returnSuccess({ file: fileDoc });
+    return;
+  }
+
+  const fileMetadata = await s3.getObjectMetadata(fileDoc.s3Path || "");
+  // Catch if file metadata does not exist. This should never happen unless something goes wonderfully wrong
+  if (!fileMetadata) {
+    FileLogic.deleteOne(fileDoc.ref);
```
Comment (laura-wenning) [inline src/pages/api/files/validate-upload.ts:30]: Rename to delete instead of deleteOne

File: src/pages/api/get-upload-url.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import "reflect-metadata";
```
Comment (laura-wenning) [inline src/pages/api/get-upload-url.ts:1]: Delete this file

File: src/pages/placeholder/file-upload.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import { Button } from "@owl-factory/components/button";
```
Comment (laura-wenning) [inline src/pages/placeholder/file-upload.tsx:1]: Move to /dev/file-upload instead of /placeholder/...

File: src/types/documents/File.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { AssetSource } from "types/enums/assetSource";
+import { BaseDocument } from "./BaseDocument";
+
+export const fileExtentionMap: Record<string, string> = {
+  "image/jpeg": "jpg",
+  "image/png": "png",
+  "image/webp": "webp",
+};
+
+export interface FileDocument extends BaseDocument {
```
Comment (laura-wenning) [inline src/types/documents/File.ts:10]: Need way to differentiate what kind of file it is, eg image, PDF, etc
Also need a creation method (see AssetSource, but also rename)

File: src/types/security/permissions.ts [hunk 10]
```diff
@@ -26,12 +26,16 @@ export const permissions: Permission[] = [
...
+  { key: "uploadFile", name: "Upload a File", collection: Collection.Files },
+
   { key: "createExternalImage", name: "Create Externally Linked Image", collection: Collection.Images },
   { key: "deleteImage", name: "Delete My Image", collection: Collection.Images },
   { key: "viewAnyImage", name: "View Any Image", collection: Collection.Images },
   { key: "searchImagesByUser", name: "Search Images By User", collection: Collection.Images },
   { key: "searchMyImages", name: "Search My Images", collection: Collection.Images },
 
+
+
```
Comment (laura-wenning) [inline src/types/security/permissions.ts:38]: Remove excess whitespace

File: src/utilities/files.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Generates a key
+ * @returns The randomly generated S3 key
+ */
+export function generateS3Key() {
+  return uuid();
+}
+
+export function generateS3Filepath(account: UserDocument, fileDoc: Partial<FileDocument>) {
```
Comment (laura-wenning) [inline src/utilities/files.ts:12]: Comment

File: src/utilities/image-upload.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { getSession } from "@owl-factory/auth/session";
```
Comment (laura-wenning) [inline src/utilities/image-upload.ts:1]: Delete this file

File: src/utilities/validation/file.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import { FileDocument } from "types/documents";
+
+export function validateFileUploadDoc(doc: Partial<FileDocument>) {
```
Comment (laura-wenning) [inline src/utilities/validation/file.ts:3]: Complete and comment these

**Comment by laura-wenning** [comment]
We also need to save the src of the file. Because we didn't. Like a dummy

**Review by lucyawrey — APPROVED** [review]
Looks good 🙂

**Pull request merged**